Authenticate network messages

pgrange opened the pull request:
<!-- Describe your change here -->

Fixes #727

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

v0d1ch committed:
Add SignableRepresentation for Message type

ffakenz committed:
Stub out a new test case for authenticated messages

ffakenz committed:
Draft withAuthentication fn

pgrange committed:
Test passes

pgrange committed:
WIP

ffakenz committed:
Introduce signatures for Authenticated messages

pgrange committed:
Working authentication network layer

v0d1ch committed:
Add new Authenticate module

v0d1ch committed:
Use the new Authenticate module in the hydra-node executable

- Provide SignableRepresentation instance for Heartbeat

Use the new Authenticate module in the hydra-node executable

- Provide SignableRepresentation instance for Heartbeat

ffakenz committed:
Add draft docs the authenticate network module

ffakenz committed:
Reorder instances and move orphans to the end of the module

ffakenz committed:
Update CHANGELOG

pgrange committed:
Implement a rountrip cbor encoding test for Authenticate type

v0d1ch committed:
CheckPoint

ffakenz committed:
Add sender party to authenticated messages

Used to check the sender is the signer.

ghost commented:
Ran some quick benchmarks 

For authenticated messages:

```
Confirmed txs: 900
Average confirmation time (ms): 172.994095922
P99: 254.08549761ms
P95: 239.86288929999995ms
P50: 165.64742ms
Invalid txs: 0
```

On master

```
Confirmed txs: 900
Average confirmation time (ms): 172.274292194
P99: 252.98727588999998ms
P95: 240.6767249ms
P50: 166.4720585ms
Invalid txs: 0
```

pgrange committed:
Refactor: introduce AuthenticatedMessage and SignedMessage

An AtuhenticatedMessager is a SignedMessage which we authenticated.

v0d1ch committed:
Add a note related to signing network messages

pgrange committed:
We don't use the party in a ReqTx

pgrange committed:
No party inside the message

This ensure that one can not forge a message from another party
and signed it with it's own key. There is no party in a message
so the party is always the signer of the Authenticated message.

abailly committed:
Failing test for logging dropped messages

ffakenz committed:
Log dropped messages

abailly committed:
Correctly set the sender's party in mock network

abailly committed:
Refresh golden files for Message

abailly committed:
Update (part of) the logs' schema

abailly committed:
Fix party in mock network used in Model

abailly committed:
Minor fix to changelog (spelling)

abailly committed:
Fix spurious import

abailly committed:
Add missing golden file for Message
